clients/src/main/java/org/apache/kafka/common/requests/BrokerRegistrationRequest.java

```diff
@@ -45,7 +46,21 @@ public short oldestAllowedVersion() {
 
         @Override
         public BrokerRegistrationRequest build(short version) {
-            return new BrokerRegistrationRequest(data, version);
+            if (version < 4) {
+                // Workaround for KAFKA-17011: for BrokerRegistrationRequest versions older than 4,
+                // translate minSupportedVersion = 0 to minSupportedVersion = 1.
+                BrokerRegistrationRequestData newData = data.duplicate();
+                for (Iterator<BrokerRegistrationRequestData.Feature> iter = newData.features().iterator();
+                     iter.hasNext(); ) {
+                    BrokerRegistrationRequestData.Feature feature = iter.next();
+                    if (feature.minSupportedVersion() == 0) {
+                        feature.setMinSupportedVersion((short) 1);
```

Comment:
There is another issue caused by changing the version from 0 to 1. #15685 add the check which assumes the "default version" is 0 

https://github.com/apache/kafka/blob/3.8/metadata/src/main/java/org/apache/kafka/controller/ClusterControlManager.java#L14

Hence, the features having min=1 can NOT pass the check. see following error message:
```
[2024-09-05 11:22:22,146] INFO [QuorumController id=2] registerBroker: event failed with UnsupportedVersionException in 127 microseconds. Exception message: Unable to register because the broker does not support version 0 of group.version. It wants a version between 1 and 1, inclusive. (org.apache.kafka.controller.QuorumController)
```

Reply:
@cmccabe @junrao @dajac FYI

Reply:
@chia7712 Thanks. Is there a way to reproduce this (e.g. an existing test)?

Reply:
>  Is there a way to reproduce this (e.g. an existing test)?

run the following command with #17084

```
TC_PATHS="tests/kafkatest/tests/core/kraft_upgrade_test.py::TestKRaftUpgrade.test_combined_mode_upgrade" \
/bin/bash tests/docker/run_tests.sh
```

Reply:
Originally I thought that we could not set version 0 with the storage or update tools since it would treat it as disabling the feature. But @dajac pointed out to me that this is happening by setting the MV and that picks default features.

However, the min version 1 should only be the case for older request versions. I thought we fixed this to so we wouldn't have the requirement for version 4 and for 4.0 where group version is introduced.

Reply:
This bug is pretty nasty because the controller side logic is in 3.8. In this case, the issue is that a broker using a newer version of the software (dev) sends its features to an old controller (3.8). Hence the features are changed to start from version 1 (fix in this PR). As the controller does not know the feature, it defaults to version 0 and requires the feature range to support it. It fails because the received range is 1-1. I think that we got the logic on the controller side wrong. At least, we should have adapted it based on this PR.

Reply:
@chia7712 This will also break with 3.9. Should we raise a blocker?

Reply:
I don't fully understand why we changed the BrokerRegistration request. I reverted the change locally and it works.

Reply:
leave short response as I'm on traveling :smile: 

> This will also break with 3.9. Should we raise a blocker?

I will raise a jira later 

> I don't fully understand why we changed the BrokerRegistration request. I reverted the change locally and it works.

see https://github.com/apache/kafka/pull/16420#issuecomment-2192619405

In short, it prevents v0 bug caused by old controller in zk migration

Reply:
> This will also break with 3.9. Should we raise a blocker?

https://issues.apache.org/jira/browse/KAFKA-17492

Reply:
I will take a deeper look into this. Thanks for the summary @dajac

Reply:
One option we can take is to release a patch release for 3.8 with a fix -- it is not ideal but it may be the simplest thing to do. In the release notes for 3.9, we may have to say for isolated mode, the controllers need to be upgraded first? And combined there is not much we can do until the patch? (Since when we shut the controller broker down, the controller will always move to an older one)


The alternative would be to do some sort of weird version negotiation -- the new broker would have to know it is the old controller and not send the feature at all, then have to share its supported range once the controller is upgraded.

Reply:
> One option we can take is to release a patch release for 3.8 with a fix -- it is not ideal but it may be the simplest thing to do. In the release notes for 3.9, we may have to say for isolated mode, the controllers need to be upgraded first? And combined there is not much we can do until the patch? (Since when we shut the controller broker down, the controller will always move to an older one)

agree. it is the simplest approach.  Also, we should add notes to 3.8 to remind users "3.8" controller can NOT work with 3.9+ broker. The error message of "unsupported version" should include this discussion too.  

> The alternative would be to do some sort of weird version negotiation -- the new broker would have to know it is the old controller and not send the feature at all, then have to share its supported range once the controller is upgraded.

Or we can take 3.8 release back like 3.3.0 ... ?

Reply:
> Or we can take 3.8 release back like 3.3.0 ... ?
Did we take the release back or did we just release 3.3.1 quickly after?

Reply:
> Did we take the release back or did we just release 3.3.1 quickly after?

It seems both are true. in the release page the 3.3.0 is gone. BUT, I'm not sure the "valid time" to take a release back

Reply:
@chia7712 : Thanks for catching this issue.

@jolshan : Regarding the fix, perhaps we could change the logic in `ClusterControlManager.java` back to the one in 3.7 like the following? This will also help prevent a potential issue in upgrading from 3.9 to a future release that increases the min supported version. 

```
        Optional<Short> finalized = finalizedFeatures.get(feature.name());
        if (finalized.isPresent()) {
            if (!VersionRange.of(feature.minSupportedVersion(), feature.maxSupportedVersion()).contains(finalized.get())) {
```

To support upgrading from 3.8.0 to 3.9, we could potentially change the implementation of ApiVersionsResponse (<4) and BrokerRegistrationRequest (<4) to skip features with supported minVersion of 0 instead of replacing 0 with 1. I argued for the latter, but the benefit from that is probably not worth the trouble with the upgrade.

Reply:
> ApiVersionsResponse (<4) and BrokerRegistrationRequest (<4) to skip features with supported minVersion of 0 instead of replacing 0 with 1. I argued for the latter, but the benefit from that is probably not worth the trouble with the upgrade.

With this approach, we don't need to fix the `ClusterControlManager.java` for 3.8, right?

Reply:
> This will also help prevent a potential issue in upgrading from 3.9 to a future release that increases the min supported version.

3.9+ should be able to handle min=0 and they are using >4 BrokerRegistrationRequest, so not sure why we need to remove this check from 3.9

Reply:
> 3.9+ should be able to handle min=0 and they are using >4 BrokerRegistrationRequest, so not sure why we need to remove this check from 3.9

I was thinking that if a future release increases the min supported version to be higher than 0, then upgrading from 3.9 will have the same issue.

Reply:
> I was thinking that if a future release increases the min supported version to be higher than 0, then upgrading from 3.9 will have the same issue.

you are right!

Thanks for all discussion. I think the following fixes should be shipped to 3.9 and trunk. 

1. **make 3.9+ broker be able to register to 3.8 controller**
skip features with minVersion of 0 instead of replacing 0 with 1 when BrokerRegistrationRequest < 4
2. **make 3.9+ controller be able to accept unknown features from future broker**
skip check of unknown features. see following sample code
```java
   Optional<Short> finalized = finalizedFeatures.get(feature.name());
        if (finalized.isPresent()) {
            if (!VersionRange.of(feature.minSupportedVersion(), feature.maxSupportedVersion()).contains(finalized.get())) {
```
We will prepare the PR this weekend

Reply:
> I was thinking that if a future release increases the min supported version to be higher than 0, then upgrading from 3.9 will have the same issue.

have second thoughts, the features of future release having min=1 should NOT join as the features can NOT be disabled. Otherwise, the broker cluster will run with inconsistent features

Reply:
> I don't fully understand why we changed the BrokerRegistration request. I reverted the change locally and it works.

The jira (https://issues.apache.org/jira/browse/KAFKA-17497) will add a e2e case to re-produce the bug.

Reply:
> have second thoughts, the features of future release having min=1 should NOT join as the features can NOT be disabled. Otherwise, the broker cluster will run with inconsistent features

Good question. @cmccabe : If we increase the minSupportedVersion of a feature in the future, could we still disable the feature? In the current code, it seems that we won't allow setting the feature to 0 if minSupportedVersion is > 1.

Reply:
When I last talked to Colin about this, I believe he said we could not disable the feature if the min version is > 0.

Reply:
> When I last talked to Colin about this, I believe he said we could not disable the feature if the min version is > 0.

@cmccabe : Could you confirm if this is case? How would people upgrade from an old release where a feature is not turned on?

Reply:
I would suspect you have to upgrade the feature before the upgrade to the new release, but we can let Colin confirm.

Reply:
Checked with Colin offline. To summarize, if we do increase minVersion in the future, we need to bump up the default version for that feature. An old release may need to first upgrade to a bridge release where the new default version could be set. Given that, we don't need to change the logic in ClusterControlManager.

Reply:
> Given that, we don't need to change the logic in ClusterControlManager.

I will sync it to KAFKA-17429 and PR#17128